eth/downloader/downloader.go

```diff
@@ -1332,27 +1352,23 @@ func (d *Downloader) processHeaders(origin uint64, td *big.Int) error {
 					if chunk[len(chunk)-1].Number.Uint64()+uint64(fsHeaderForceVerify) > pivot {
 						frequency = 1
 					}
-					if n, err := d.insertHeaders(chunk, frequency); err != nil {
+					res := d.lightchain.InsertHeaderChain(chunk, frequency)
+					// TODO(whilei): again, send error to events
+					if res.Error != nil {
 						// If some headers were inserted, add them too to the rollback list
-						if n > 0 {
-							rollback = append(rollback, chunk[:n]...)
+						if res.Index > 0 {
```

Comment:
So here's how and why the core `Insert` functions return a `0` index on successful import; but still just an idiom, b/c this could as easily be `if res.Index < len(chunk)-1`.